src/Modules/Trinity.FFI/Trinity.FFI.PythonModule/collect.py

```diff
@@ -1,60 +1,83 @@
+                       default=False, nargs='?',
+                       const=True)
 
 
-@curry
-def copy_to(_to, _from):
-    """shutils sucks, it failed frequently in *NIX OS.
+def dir_at(*args, create_dir=True, then_call=None):
     """
-    _to = os.path.join(_to, os.path.split(_from)[-1])
-    print(_to)
-    with open(_from, 'rb') as _from_file, open(_to, 'wb') as _to_file:
-        and_then(
-            _from_file.read,
-            _to_file.write
-        )(None)
-
-
-# build dlls
-#os.system('{} "{}"'.format(BUILD_SCRIPT_CMD, BUILD_SCRIPT_PATH))
-
-def recur_listdir(dir):
-    filenames = os.listdir(dir)
-    for filename in map(lambda _: os.path.join(dir, _), filenames):
-        if os.path.isdir(filename):
-            yield from recur_listdir(filename)
-        else:
-            yield filename
-
-
-# copy dlls
-for dll_path in (CORECLR_PATH, '../TslAssembly'): 
-    (linq.Flow(dll_path)
-     .Then(
-        and_then(recur_listdir))
-     .Filter(lambda _: _.endswith('.dll') or _.endswith('.pyd') or _.endswith('.py'))
-     .Each(copy_to(os.path.join(CURRENT_DIR, 'GraphEngine', 'ffi'))))
-
-
-# copy exe
-(linq.Flow(os.path.join(CORECLR_PATH, '../'))
-     .Then(
-        and_then(recur_listdir))
-     .Filter(lambda _: _.endswith('.exe'))
-     .Each(copy_to(os.path.join(CURRENT_DIR, 'GraphEngine', 'Command'))))
+    join the path and try to create the directory.
+    """
+    path = os.path.join(*args)
+    if create_dir and not os.path.exists(path):
+        try:
+            os.mkdir(path)
+        except FileExistsError:
+            pass
+    if then_call is not None:
+        return then_call(path)
+    return path
+
+
+if is_windows:
+    # build dlls
+    if cmdparser.parse_args().nocache:
+        os.system('{} "{}"'.format(BUILD_SCRIPT_CMD, BUILD_SCRIPT_PATH))
+
+    to_pymodule_dir = lambda path, with_postfix, to_module: (
+        linq.Flow(path)
+            .Then(recur_listdir)
+            .Filter(lambda _: any(_.endswith(postfix) for postfix in with_postfix))
+            .Each(dir_at(CURRENT_DIR, 'GraphEngine', to_module,
+                         then_call=copy_to)))
+    # copy dlls
+
+    to_pymodule_dir(CORECLR_PATH,
+                    with_postfix=['.py', '.dll', '.pyd'],
+                    to_module='ffi')
+
+    # build Trinity.StorageVersionController.csproj
+    dir_at(pardir_of(pardir_of(BUILD_SCRIPT_PATH)), 'DynamicLoading',
+           then_call=lambda _: os.system('cd {} && dotnet restore && dotnet build'.format(_)))
+
+    to_pymodule_dir('../../DynamicLoading',
+                    with_postfix=['.dll'],
+                    to_module='ffi')
+
+    # copy exe
+    to_pymodule_dir(pardir_of(CORECLR_PATH),
+                    with_postfix=['.exe'],
+                    to_module='Command')
+
+    # add newtonsoft.json.dll
+    # TODO: a robust way to find out newtonsoft.json.dll.
+    to_pymodule_dir(r'C:\Program Files\dotnet\sdk\NuGetFallbackFolder\newtonsoft.json\9.0.1\lib\netstandard1.0',
```

Comment:
if we assume 'dotnet' to be an available command, we can use it to query the location of the package cache.